### PR TITLE
Kotlin multiplatform client

### PR DESCRIPTION
**Pull request by andrewemery**
Support for the generation of [Kotlin Multiplatform](https://kotlinlang.org/docs/reference/multiplatform.html) clients.

**Background**

The ```kotlin``` language defined in the Swagger Codegen project currently generates a Kotlin client that can run on the JVM. Somewhat recently, JetBrains introduced [Kotlin Multiplatform](https://kotlinlang.org/docs/reference/multiplatform.html), a platform that provides the ability to share code between various platforms: JVM, JS, iOS, Windows, etc. This pull request adds the ability to generate a Kotlin Multiplatform client.

**Library**

The pull request includes Kotlin Multiplatform generation under the ```kotlin``` language and introduces two library options:
1. ```jvm```: (default) to target the existing Kotlin JVM generation.
2. ```multiplatform```: to target Kotlin Multiplatform generation.

**Limitations**

The current pull request only currently adds support for JVM and iOS platforms but support for other platforms should be trivial to add.

At present, ```date``` and ```date-time``` objects are rendered as strings. No attempt is made to integrate any third-party date/time libraries.

**Tests**

The current implementation doesn't include any tests beyond those already included as part of the ```kotlin``` language implementation. A test project can be found [here](https://github.com/andrewemery/openapi-generator-kotlin-multiplatform-petstore) that generates the ```petstore``` sample and includes tests to verify the correctness of the implementation.

Comments welcome.

**References**

https://github.com/OpenAPITools/openapi-generator/issues/3899
https://github.com/andrewemery/openapi-generator-kotlin-multiplatform-petstore

### CODE REVIEW

File: bin/windows/kotlin-client-petstore-multiplatform.bat [hunk 1]
```diff
@@ -0,0 +1,10 @@
+set executable=.\modules\openapi-generator-cli\target\openapi-generator-cli.jar
+
+If Not Exist %executable% (
+  mvn clean package
+)
+
+REM set JAVA_OPTS=%JAVA_OPTS% -Xmx1024M -DloggerPath=conf/log4j.properties
+set ags=generate  --artifact-id "kotlin-client-petstore-multiplatform" -i modules\swagger-codegen\src\test\resources\2_0\petstore.yaml -g kotlin --library multiplatform -o samples\client\petstore\kotlin-multiplatform
```
Comment (wing328) [inline bin/windows/kotlin-client-petstore-multiplatform.bat]: For `-i`, please use `modules\openapi-generator\src\test\resources\2_0\petstore.yaml` instead

Comment (andrewemery) [reply bin/windows/kotlin-client-petstore-multiplatform.bat]: Thanks. Updated.

**Comment by wing328** [comment]
Please run the following to update the documentation so as to address the CircleCI failure:
```
"./bin/utils/export_docs_generators.sh"
"./bin/utils/copy-to-website.sh"
"./bin/utils/export_generators_readme.sh"
```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java [hunk 2]
```diff
@@ -151,25 +241,97 @@ public void processOpts() {
             typeMapping.put("list", "kotlin.collections.List");
             additionalProperties.put("isList", true);
         }
+    }
+
+    @Override
+    public Map<String, Object> postProcessModels(Map<String, Object> objs) {
+
+        // workaround for: https://github.com/swagger-api/swagger-codegen/issues/4258
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: Do we still need this workaround? 

I think the issue has been fixed in OpenAPI Generator

Comment (andrewemery) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: Not sure. I couldn't find a reference to the issue in OpenAPI Generator (and I couldn't remember in what scenario it was necessary). I'll run through some tests and get back to you.

Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: Yes, please. If it's still an issue, we need to track it and get it fixed.

Comment (andrewemery) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: Confirmed fixed. Unnecessary code removed.

Comment (andrewemery) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: I've also included a fix for [this](https://github.com/OpenAPITools/openapi-generator/issues/3917) issue. Happy to submit as a separate pull request if desired.

https://github.com/OpenAPITools/openapi-generator/issues/3917

Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: Prefer a separate PR

Comment (andrewemery) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: Separate pull request created:
https://github.com/OpenAPITools/openapi-generator/pull/3928

**Comment by wing328** [comment]
I got the following errors when testing the multiplatform client locally:

```
➜  kotlin-multiplatform git:(andrewemery-kotlin-multiplatform-client) ✗ gradle build
Starting a Gradle Daemon (subsequent builds will be faster)

> Configure project :
Kotlin Multiplatform Projects are an experimental feature.

Please wait while Kotlin/Native compiler 1.3.50 is being installed.
Download https://download.jetbrains.com/kotlin/native/builds/releases/1.3.50/macos/kotlin-native-macos-1.3.50.tar.gz
Download https://download.jetbrains.com/kotlin/native/builds/releases/1.3.50/macos/kotlin-native-macos-1.3.50.tar.gz finished, took 30 s 697 ms
Unpack Kotlin/Native compiler to /Users/williamcheng/.konan/kotlin-native-macos-1.3.50
Unpack Kotlin/Native compiler to /Users/williamcheng/.konan/kotlin-native-macos-1.3.50 finished, took 28 s 111 ms

> Task :compileKotlinIosArm64 FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlinIosArm64'.
> Could not resolve all files for configuration ':iosArm64CompileKlibraries'.
   > Could not resolve org.jetbrains.kotlinx:kotlinx-coroutines-core-native:1.3.1.
     Required by:
         project :
      > Could not resolve org.jetbrains.kotlinx:kotlinx-coroutines-core-native:1.3.1.
         > Could not parse module metadata https://jcenter.bintray.com/org/jetbrains/kotlinx/kotlinx-coroutines-core-native/1.3.1/kotlinx-coroutines-core-native-1.3.1.module
            > Unsupported format version '1.0' specified in module metadata. This version of Gradle supports format version 0.4 only.
   > Could not resolve org.jetbrains.kotlinx:kotlinx-coroutines-core-native:1.3.1.
     Required by:
         project : > io.ktor:ktor-client-ios-iosarm64:1.2.4
         project : > io.ktor:ktor-client-json-iosarm64:1.2.4
         project : > io.ktor:ktor-client-serialization-iosarm64:1.2.4
         project : > io.ktor:ktor-client-ios-iosarm64:1.2.4 > io.ktor:ktor-client-core-native:1.2.4 > io.ktor:ktor-client-core-iosarm64:1.2.4
         project : > io.ktor:ktor-client-ios-iosarm64:1.2.4 > org.jetbrains.kotlinx:kotlinx-coroutines-io-native:0.1.14 > org.jetbrains.kotlinx:kotlinx-coroutines-io-iosarm64:0.1.14
         project : > io.ktor:ktor-client-ios-iosarm64:1.2.4 > io.ktor:ktor-client-core-native:1.2.4 > io.ktor:ktor-client-core-iosarm64:1.2.4 > io.ktor:ktor-http-native:1.2.4 > io.ktor:ktor-http-iosarm64:1.2.4
         project : > io.ktor:ktor-client-ios-iosarm64:1.2.4 > io.ktor:ktor-client-core-native:1.2.4 > io.ktor:ktor-client-core-iosarm64:1.2.4 > io.ktor:ktor-http-cio-native:1.2.4 > io.ktor:ktor-http-cio-iosarm64:1.2.4
         project : > io.ktor:ktor-client-ios-iosarm64:1.2.4 > io.ktor:ktor-client-core-native:1.2.4 > io.ktor:ktor-client-core-iosarm64:1.2.4 > io.ktor:ktor-http-native:1.2.4 > io.ktor:ktor-http-iosarm64:1.2.4 > io.ktor:ktor-utils-native:1.2.4 > io.ktor:ktor-utils-iosarm64:1.2.4
      > Could not resolve org.jetbrains.kotlinx:kotlinx-coroutines-core-native:1.3.1.
         > Could not parse module metadata https://jcenter.bintray.com/org/jetbrains/kotlinx/kotlinx-coroutines-core-native/1.3.1/kotlinx-coroutines-core-native-1.3.1.module
            > Unsupported format version '1.0' specified in module metadata. This version of Gradle supports format version 0.4 only.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2m 11s
1 actionable task: 1 executed

```

Is it supposed to work on Mac as well?

**Comment by andrewemery** [comment]
> I got the following errors when testing the multiplatform client locally:

Try it with the wrapper:
```
./samples/client/petstore/kotlin-multiplatform/gradlew assemble
```

> Is it supposed to work on Mac as well?

Yes. I've tested on Mac and Windows.

**Comment by wing328** [comment]
Thanks for the tips. It works for me in my machine:
```
CONST Null type=kotlin.Nothing? value=null

> Task :compileKotlinMetadata
w: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (124, 15): Unnecessary safe call on a non-null receiver of type Array<String>
w: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (169, 13): Unnecessary safe call on a non-null receiver of type Array<String>
w: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (314, 119): This declaration is experimental and its usage should be marked with '@io.ktor.util.KtorExperimentalAPI' or '@UseExperimental(io.ktor.util.KtorExperimentalAPI::class)'
w: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt: (239, 17): Unnecessary safe call on a non-null receiver of type String
w: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt: (241, 17): Unnecessary safe call on a non-null receiver of type String

BUILD SUCCESSFUL in 10m 1s
13 actionable tasks: 13 executed
```

File: samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            
+            
+            
+            
+            
+            
+            
+            
+            
+
```
Comment (wing328) [inline samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt]: Too many empty lines. I'll see what I can do to fix these in a separate PR

**Comment by wing328** [comment]
Tweet: 

Hacker News: https://news.ycombinator.com/item?id=21089885

Reddit: https://www.reddit.com/r/Kotlin/comments/d9xndy/openapi_generators_kotlin_client_generator_now/

**Comment by andrewemery** [comment]
Thanks. Looking forward to community feedback and improvements.

**Pull request merged**